### ISSUE TITLE
`ShaderSource::Glsl { defines }` is underdocumented

### ISSUE THREAD

**Issue by kpreid** [gfx-rs/wgpu#2964]
The documentation of `ShaderSource::Glsl` says:

> <code>defines: [FastHashMap](https://docs.rs/naga/0.9.0/x86_64-unknown-linux-gnu/naga/type.FastHashMap.html)&lt;[String](https://doc.rust-lang.org/nightly/alloc/string/struct.String.html), [String](https://doc.rust-lang.org/nightly/alloc/string/struct.String.html)></code>
> Defines to unlock configured shader features.

This only vaguely describes the purpose (What is a “feature”? How does a shader have them? What does “unlock” mean, formally?) but it does not tell the user what format of input to give it, or what to do if they are not using this functionality.

I would assume this does something equivalent to prepending `format!("#define {key} {value}\n")` to the shader source, but if so, it should say that — there's no need to be abstract, since we're talking about the specific language GLSL.

**Comment by tombh** [comment]
I agree. But just to note, I managed to successfully inject the following:
```glsl
#define fragCoord gl_FragCoord.xy
```

Using:
```rust
let mut defines: wgpu::naga::FastHashMap<String, String> = Default::default();
defines.insert("fragCoord".into(), "gl_FragCoord.xy".into());
```
